packages/agent/shared/agent.ts

```diff
@@ -1,74 +1,48 @@
-import { Operation, resource, spawn } from 'effection';
+import { createStream, Task, Resource, spawn, ensure } from 'effection';
 import { on, once } from '@effection/events';
-import { ChainableSubscribable, createSubscription } from '@effection/subscription';
-import { AgentProtocol, AgentEvent, Command, Run } from './protocol';
+import { AgentProtocol, AgentEvent, Command } from './protocol';
 
 export * from './protocol';
 
 export interface AgentOptions {
   agentId?: string;
-  data: unknown;
-  createSocket: () => Socket;
+  data?: Record<string, unknown>;
 }
 
-export class Agent implements AgentProtocol {
-  constructor(private socket: Socket & EventTarget, private options: AgentOptions) {}
+export function createAgent(socket: Socket & EventTarget, options: AgentOptions): Resource<AgentProtocol> {
+  return {
+    *init(scope: Task) {
+      let subscription = createStream<Command, CloseEvent>((publish) => function*(task) {
+        task.run(on(socket, 'message')
+          .map((event) => event as MessageEvent)
+          .map(event => JSON.parse(event.data) as Command)
+          .forEach(publish));
+
+        let closed = yield once(socket, 'close');
+        return closed;
+      }).subscribe(scope);
```

Comment:
I believe `Stream` is a resource that produces a subscription, so you can just yield to it instead.

```suggestion
      let subscription = yield createStream<Command, CloseEvent>((publish) => function*(task) {
        task.run(on(socket, 'message')
          .map((event) => event as MessageEvent)
          .map(event => JSON.parse(event.data) as Command)
          .forEach(publish));

        let closed = yield once(socket, 'close');
        return closed;
      });
```